### DISCUSSION TITLE
Make OpenSSL use pointers from input buffer for holding CMS structure instead of allocating memory

### DISCUSSION THREAD

**Discussion by mura1i** [openssl/openssl#24822] [Q&A] [answered]
An embedded device with limited memory is already using OpenSSL for communication.

I am implementing a CMS signature verification wrapper library using OpenSSL that should work for the device.

We already have the below code which is working.

```c
CMS_ContentInfo *cms = d2i_CMS_ContentInfo(nullptr, in, len); // allocates memory 1

if (CMS_verify(cms, nullptr, store, nullptr, nullptr, 0))
{
    ASN1_OCTET_STRING **data = CMS_get0_content(cms);
    len = (*data)->length);
    out = OPENSSL_zalloc(len); // allocates memory 2
    memcpy(out, (*data)->data, len);
}

CMS_ContentInfo_free(cms); // frees memory 1
```

But it allocates redundant memory 2 times.

Any way to make OpenSSL use the input buffer itself for holding the data?

**Comment by mattcaswell** [comment]
```` c
CMS_ContentInfo *cms = d2i_CMS_ContentInfo(nullptr, in, len); // allocates memory 1
````
I don't think there is anyway to avoid this allocation

```` c
ASN1_OCTET_STRING **data = CMS_get0_content(cms);
out = OPENSSL_zalloc(len); // allocates memory 2
memcpy(out, (*data)->data, len);
````

This allocation looks quite strange to me. You seem to be copying the content string data...but using the length of the **input** buffer instead of the length of the content. This would have made more sense to me:

```` c
ASN1_OCTET_STRING **data = CMS_get0_content(cms);
out = OPENSSL_zalloc((*data)->length); // allocates memory 2
memcpy(out, (*data)->data, (*data)->length);
````
With that amended form, the copied data can go anywhere you like. The input buffer is no longer being used by the `cms` object, so there's no reason you couldn't copy it back into the input buffer if you so wish:

```` c
ASN1_OCTET_STRING **data = CMS_get0_content(cms);
assert((*data)->length <= len); // We would always expect this to be true
memcpy(in, (*data)->data, (*data)->length);
````

**Reply by mura1i** [reply]
Thank you @mattcaswell for the reply 😃

But we are not allowed to modify the input buffer. 😞

> This allocation looks quite strange to me.

The second allocation does look strange because I had missed one line in the problem statement 😕

```c
len = (*data)->length);
```

We are doing this since we need to output the length also to the caller. 

Can I modify the question with this updated info?

**Reply by mattcaswell** [reply]
> Can I modify the question with this updated info?

Sure

**Reply by mattcaswell** [reply, marked as answer]
> But we are not allowed to modify the input buffer. 

Your only other option is to keep the CMS object around for a while and reference `data->data` and `data->length` directly without taking a copy of it and allocating a new buffer for it. The buffer will remain valid for as long as the CMS object exists and you haven't modified it.

**Reply by mura1i** [reply]
Ok. Thank you for the suggestion. That will reduce the memory footprint ~33%.

Can I raise an issue from this query?

**Reply by mattcaswell** [reply]
> Can I raise an issue from this query?

Issues should be for reporting bugs, documentation issues or feature requests. I haven't heard you talk about one of those things in this discussion. But if you have one of those things then please feel free to raise an issue for it.

**Reply by mura1i** [reply]
Doesn't _"Make OpenSSL use pointers from input buffer for holding CMS structure instead of allocating memory"_ qualify as a feature request?

**Reply by mattcaswell** [reply]
I took that as a question for how to achieve that, rather than a feature request for it. Yes, that would qualify as a feature request - although not one I think is technically feasible.

**Reply by mura1i** [reply]
Will raise it hoping someone might find a way 🤞 (even though it won't be useful for me)
